coding.go

```diff
@@ -22,38 +25,84 @@ const _ = pb.DoNotUpgradeFileEverItWillChangeYourHashes
 
 // unmarshal decodes raw data into a *Node instance.
 // The conversion uses an intermediate PBNode.
-func (n *ProtoNode) unmarshal(encoded []byte) error {
-	var pbn pb.PBNode
-	if err := pbn.Unmarshal(encoded); err != nil {
-		return fmt.Errorf("unmarshal failed. %v", err)
+func unmarshal(encodedBytes []byte) (*ProtoNode, error) {
+	nb := dagpb.Type.PBNode.NewBuilder()
+	if err := dagpb.DecodeBytes(nb, encodedBytes); err != nil {
+		return nil, err
 	}
+	nd := nb.Build()
+	return fromImmutableNode(&immutableProtoNode{encodedBytes, nd.(dagpb.PBNode)}), nil
+}
 
-	pbnl := pbn.GetLinks()
-	n.links = make([]*ipld.Link, len(pbnl))
-	for i, l := range pbnl {
-		n.links[i] = &ipld.Link{Name: l.GetName(), Size: l.GetTsize()}
-		c, err := cid.Cast(l.GetHash())
-		if err != nil {
-			return fmt.Errorf("link hash #%d is not valid multihash. %v", i, err)
+func fromImmutableNode(encoded *immutableProtoNode) *ProtoNode {
+	n := new(ProtoNode)
+	n.encoded = encoded
+	if n.encoded.PBNode.Data.Exists() {
+		n.data = n.encoded.PBNode.Data.Must().Bytes()
+	}
+	numLinks := n.encoded.PBNode.Links.Length()
+	n.links = make([]*format.Link, numLinks)
+	linkAllocs := make([]format.Link, numLinks)
+	for i := int64(0); i < numLinks; i++ {
+		next := n.encoded.PBNode.Links.Lookup(i)
+		name := ""
+		if next.FieldName().Exists() {
+			name = next.FieldName().Must().String()
 		}
-		n.links[i].Cid = c
+		c := cid.Undef
+		c = next.FieldHash().Link().(cidlink.Link).Cid
```

Comment:
honestly, I consider "second link interface" far enough off not to care that much. Per @warpfork for the time being it's fine to just do casts like this.

Reply:
Yeah.  I consider this fine until we actually seriously expect existence of another possible inhabitant.  We don't, today; and if we ever add one, then we will just have to make a point to walk through all these repos and add real branches for this (and it should be easy to do by just grepping for casts).

(Big picture? I'm often regretting the `Link` interface at all.  The hope I had long ago was that that interface would help us discover what bits of the CID type _really matter_ vs what are historically accumulated noise, and also keep go-cid's transitive dep sprawl from going everywhere all the time.  Those goals still don't seem wrong, and may have even been partially successful... but it's certainly dubious whether that those discoveries were gained cost-effectively, and we still end up casting out to a concrete form all the time because the interface still doesn't describe all the functionality we regularly need.  Code in `go-ipld-prime` often manages to work using interface alone (success), but our applications tend not to (not success).  Now: moving in _any_ direction at all on this requires careful change planning, and is well _well_ beyond what we can think about bringing up into scope for any of these PRs.  But meanwhile these casts seem pretty tame to me.)